### PR TITLE
Allow for ranged ids in some commands

### PR DESCRIPTION
**Pull request by stuartskelton**
this implements #59.

this means you can now do:
todo d 1,2,3,4
todo d 1,80,9,2,5-10,11

this works for:
Prioritizing
Unprioritizing
Completing
Un-completing
Archiving
Un-Archiving

I have left Edit out, as its its not as straight forward.

### CODE REVIEW

File: todolist/app.go [hunk 1]
```diff
@@ -193,6 +193,36 @@ func (a *App) getId(input string) (int, *Todo) {
...
+func (a *App) getIds(input string) []int {
+	var ids []int
+	idGroups := strings.Split(input, ",")
+
+	singleNumberRE, _ := regexp.Compile("\\d+")
+	rangeNumberRE, _ := regexp.Compile("(\\d+)-(\\d+)")
+
+	for _, idGroup := range idGroups {
+		if matches := rangeNumberRE.FindStringSubmatch(idGroup); len(matches) > 0 {
+			lowerID, _ := strconv.Atoi(matches[1])
```
Comment (stuartskelton) [inline todolist/app.go]: this is probably not needed, as the match is "\d", but  lets play it safe.

**Comment by gammons** [comment]
oh wow, I am super excited for this.  I'll be reviewing later today!

File: todolist/app.go [hunk 2]
```diff
@@ -193,6 +193,36 @@ func (a *App) getId(input string) (int, *Todo) {
...
+			upperID, _ := strconv.Atoi(matches[2])
+			if lowerID >= upperID {
+				fmt.Printf("Invalid id group: %s.\n", idGroup)
+				break
+			}
+			for id := lowerID; id <= upperID; id++ {
+				ids = append(ids, id)
+			}
```
Comment (gammons) [inline todolist/app.go]: can you extract this block into its own function?  Maybe something like `findIdsByRange` or something similar, for better readability.

File: todolist/app_test.go [hunk 3]
```diff
@@ -43,3 +43,22 @@ func TestAddTodoWithEuropeanDates(t *testing.T) {
...
+	assert.Equal([]int{6, 10, 8, 4}, app.getIds("6,10,8,4"))
+	// a correct range
+	assert.Equal([]int{6, 7, 8}, app.getIds("6-8"))
+	// some incorrect ranges
+	assert.Equal(0, len(app.getIds("6-6")))
+	assert.Equal(0, len(app.getIds("8-6")))
+	// some compsite ranges
+	assert.Equal([]int{5, 6, 7, 8, 10, 11, 9}, app.getIds("5,6-8,10-11,9"))
+
+}
```
Comment (gammons) [inline todolist/app_test.go]: 👍 love the tests.

File: todolist/todo_list.go [hunk 4]
```diff
@@ -15,57 +15,93 @@ func (t *TodoList) Add(todo *Todo) {
...
-func (t *TodoList) Delete(id int) {
-	i := -1
-	for index, todo := range t.Data {
-		if todo.Id == id {
-			i = index
+func (t *TodoList) Delete(ids ...int) {
+	for _, id := range ids {
+		todo := t.FindById(id)
+		if todo == nil {
+			break
```
Comment (gammons) [inline todolist/todo_list.go]: looks like you're doing `continue` for the other functions.  should this be `continue` as well?

**Comment by gammons** [comment]
minor nitpicks. Otherwise this PR looks 🍺 !

File: todolist/todo_list.go (hunk 4 above)
Comment (stuartskelton) [inline todolist/todo_list.go]: this is why I was thinking about refactoring this in the same vain as the date_filter.

File: todolist/app.go (hunk 2 above)
Comment (stuartskelton) [inline todolist/app.go]: Ok I have update my PR, it might not be what you were expecting.

I am happy to revert these changes are they are a little deeper that what you may have wanted.

File: todolist/app.go [hunk 5]
```diff
@@ -174,30 +179,21 @@ func (a *App) UnprioritizeTodo(input string) {
 	fmt.Println("Todo un-prioritized.")
 }
 
-func (a *App) getId(input string) (int, *Todo) {
+func (a *App) getId(input string) int {
```
Comment (gammons) [inline todolist/app.go]: I love this refactor.  it always bothered me that `getId` would return the todo.

Comment (stuartskelton) [inline todolist/app.go]: Im glad you agree.

**Comment by gammons** [comment]
This is great work @stuartskelton I think it looks fantastic.  Are you comfortable where it is?  Or are you planning on any more refactors? (I don't think it needs any!)

**Comment by gammons** [comment]
@stuartskelton the functionality is great.  One minor nitpick - when I delete a todo range, the output from app.go is "Todo deleted.".  I think the confirmation message at the end should reflect if it's singular/plural.

**Comment by stuartskelton** [comment]
I agree, and that could be true of any action, I was going to save this for my next PR, but here is a sneaky peak.

```
func (a *App) UnprioritizeTodo(input string) {
	a.todoChanger(input, "un-prioritized", func(ids ...int) { a.TodoList.Unprioritize(ids...) })
}

func (a *App) todoChanger(input, action string, todoChanger func(ids ...int)) {
	a.Load()
	ids := a.getIds(input)
	if len(ids) == 0 {
            return
	}
	todoChanger(ids...)
	a.Save()

	fmt.Printf("Todo %s.", action)
}
````

As you can see I could cange the fmt.print line above to be something like....

```
todoOrTodos := "Todo"
if len(ids) > 1 {
	todoOrTodos = "Todos"
}
fmt.Printf("%s %s.\n", todoOrTodos, action)
```

**Comment by stuartskelton** [comment]
if your not a fan of the above, I could make a func (written on the fly)

```
func emitTodoActionMessage (action string, numOfIdsChanged int) {
    todoOrTodos := "Todo"
    if numOfIdsChanged > 1 {
        todoOrTodos = "Todos"
    }
    fmt.Printf("%s %s.\n", todoOrTodos, action)
}
```

**Comment by gammons** [comment]
@stuartskelton rails has a [`pluralize`](https://apidock.com/rails/ActionView/Helpers/TextHelper/pluralize) utility method.  Maybe we could borrow from its syntax?

**Comment by stuartskelton** [comment]
I am happy with this, Ill throw up a new PR with my other changes above, as that should be a separate piece of work.

**Comment by stuartskelton** [comment]
had to fix some conflicts but looks good.

**Comment by gammons** [comment]
nice work on this @stuartskelton thanks!
